File: leetcode/javascript/backtracking/generateParentheses(medium)/generateParentheses.js
/* https://leetcode.com/problems/generate-parentheses/description/

Учитывая n пар круглых скобок, напишите функцию, которая генерирует все комбинации правильных круглых скобок.

Input: n = 3
Output: ["((()))","(()())","(())()","()(())","()()()"]

Input: n = 1
Output: ["()"]
*/

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const result = [];

    function backtrack(s = '', left = 0, right = 0) {
        if (s.length === 2 * n) {
            result.push(s);
            return;
        }
        if (left < n) backtrack(s + '(', left + 1, right);
        if (right < left) backtrack(s + ')', left, right + 1);
    }

    backtrack();
    return result;
};

console.log(generateParenthesis(3)); // ["((()))","(()())","(())()","()(())","()()()"]

File: leetcode/go/backtracking/generateParentheses(medium)/generateParentheses.go
package main

import "fmt"

/* 22. Generate Parentheses
https://leetcode.com/problems/generate-parentheses/description/

Учитывая n пар круглых скобок, напишите функцию, которая генерирует все комбинации правильных круглых скобок.

Input: n = 3
Output: ["((()))","(()())","(())()","()(())","()()()"]

Input: n = 1
Output: ["()"]
*/

func main() {
	fmt.Println(generateParenthesis(3)) // ["((()))","(()())","(())()","()(())","()()()"]
}

// generateParenthesis генерирует все правильные скобочные последовательности длины 2n.
// time: O(4^n / sqrt(n)) - количество возможных комбинаций. space: O(n) - глубина рекурсии.
func generateParenthesis(n int) []string {
	var result []string                  // Срез для хранения результатов
	var backtrack func(string, int, int) // Внутренняя рекурсивная функция для генерации последовательностей

	backtrack = func(current string, open int, close int) {
		// Если достигли нужной длины, добавляем в результат
		if len(current) == 2*n {
			result = append(result, current)
			return
		}

		// Добавляем открывающую скобку, если еще не достигли максимума
		if open < n {
			backtrack(current+"(", open+1, close)
		}

		// Добавляем закрывающую скобку, если она соответствует открывающей
		if close < open {
			backtrack(current+")", open, close+1)
		}
	}

	backtrack("", 0, 0) // Начинаем с пустой строки и нулевых счетчиков
	return result
}
